### PR TITLE
fix(insights): make sure change in userToken can't reset the search parameters

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

previously we relied on a configure widget to set the userToken and clickAnalytics parameters. Unfortunately when the insights middleware is the first parameter, and you have a userToken set via cookie, the widget is removed and added again (setUserTokenAndSearch is called twice), and thus a change event is triggered before the other widgets are even mounted.

**Result**

Helper methods are used to prevent a change event being called by setting or updating userToken. It still triggers a new search as expected

fixes https://github.com/algolia/instantsearch.js/issues/5100

### CODE REVIEW

File: src/types/middleware.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import type { AtLeastOne } from './utils';
 export type MiddlewareDefinition<TUiState extends UiState = UiState> = {
   onStateChange(options: { uiState: TUiState }): void;
   subscribe(): void;
+  started(): void;
```
Comment (aymeric-giraudet) [inline src/types/middleware.ts:8]: So this means user-made custom middlewares now need to implement `started()` as well ?
Could it be optional ?

Comment (Haroenv) [reply src/types/middleware.ts:8]: It's actually optional in the public definition, but our own middleware use the type without optional keys @aymeric-giraudet

**Review by aymeric-giraudet — APPROVED** [review]
All good then ! 😄

**Review by FabienMotte — APPROVED** [review]

File: src/types/middleware.ts (hunk 1 above)
Comment (Haroenv) [reply src/types/middleware.ts:8]: see https://github.com/algolia/instantsearch.js/blob/6e2bf7b5cb0f56395938b17d7ef44f986ea45ca1/src/types/middleware.ts#L22 :)

**Pull request merged**